File: PayJunction VB/PayJunctionLibrary/PayJunction/PayJunctionClient.vb
Imports RestSharp
Imports RestSharp.Authenticators

Namespace PayJunction
    Class PayJunctionClient
        Private Shared EndPointBaseUrl As Dictionary(Of PaymentClientMode, String)

        Protected Property APIKey() As String
            Get
                Return m_APIKey
            End Get
            Set
                m_APIKey = Value
            End Set
        End Property
        Private m_APIKey As String
        Protected Property UserID() As String
            Get
                Return m_UserID
            End Get
            Set
                m_UserID = Value
            End Set
        End Property
        Private m_UserID As String
        Protected Property Password() As String
            Get
                Return m_Password
            End Get
            Set
                m_Password = Value
            End Set
        End Property
        Private m_Password As String
        Protected Property Mode() As PaymentClientMode
            Get
                Return m_Mode
            End Get
            Set
                m_Mode = Value
            End Set
        End Property
        Private m_Mode As PaymentClientMode

        Shared Sub New()
            EndPointBaseUrl = New Dictionary(Of PaymentClientMode, String)()
            EndPointBaseUrl.Add(PaymentClientMode.Test, "https://api.payjunctionlabs.com/")
            EndPointBaseUrl.Add(PaymentClientMode.Live, "https://api.payjunction.com/")
        End Sub

        Protected Overridable ReadOnly Property BaseEndpoint() As String
            Get
                Return EndPointBaseUrl(Mode)
            End Get
        End Property

        Public Sub New(appKey As String, apiUser As String, apiPassword As String, mode As PaymentClientMode)
            Me.APIKey = appKey
            Me.UserID = apiUser
            Me.Password = apiPassword
            Me.Mode = mode
        End Sub

        ''' <summary>
        ''' Executes request. Check request object after execution to see response.
        ''' </summary>
        ''' <param name="request">Request to process</param>
        Public Function ExecuteRequest(Of T As Response)(request As PayJunction.Request(Of T)) As Boolean
            If request.IsValid() Then
                Dim client As New RestSharp.RestClient(BaseEndpoint)
                Dim req As New RestSharp.RestRequest(request.EndPoint, request.Method)
                req.AddHeader("X-PJ-Application-Key", APIKey)
                For Each s As String In request.Headers.Keys
                    If Not String.IsNullOrWhiteSpace(request.Headers(s)) Then
                        req.AddHeader(s, request.Headers(s))
                    End If
                Next

                For Each s As String In request.Params.Keys
                    If Not String.IsNullOrWhiteSpace(request.Params(s)) Then
                        req.AddParameter(s, request.Params(s))
                    End If
                Next
                client.Authenticator = New HttpBasicAuthenticator(UserID, Password)
                Dim resp As IRestResponse = client.Execute(req)
                Console.WriteLine(vbLf & vbLf & "Raw Json Received" & vbLf)
                Console.WriteLine(resp.Content)
                Console.WriteLine(vbLf & vbLf)
                request.RawResponse = resp.Content
                request.Response = resp
                Return True
            End If

            Return False
        End Function
    End Class
End Namespace

File: PayJunction C#/ConsoleApplication2/PayJunction/PayJunctionClient.cs
using Newtonsoft.Json;
using PayJunction;
using RestSharp;
using RestSharp.Authenticators;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;

namespace PayJunction
{
    class PayJunctionClient
    {
        private static Dictionary<PaymentClientMode, string> EndPointBaseUrl;

        protected string APIKey { get; set; }
        protected string UserID { get; set; }
        protected string Password { get; set; }
        protected PaymentClientMode Mode { get; set; }

        static PayJunctionClient()
        {
            EndPointBaseUrl = new Dictionary<PaymentClientMode, string>();
            EndPointBaseUrl.Add(PaymentClientMode.Test, "https://api.payjunctionlabs.com/");
            EndPointBaseUrl.Add(PaymentClientMode.Live, "https://api.payjunction.com/");
        }

        virtual protected string BaseEndpoint
        {
            get
            {
                return EndPointBaseUrl[Mode];
            }
        }

        public PayJunctionClient(string appKey,string apiUser,string apiPassword,PaymentClientMode mode)
        {
            this.APIKey = appKey;
            this.UserID = apiUser;
            this.Password = apiPassword;
            this.Mode = mode;
        }

        /// <summary>
        /// Executes request. Check request object after execution to see response.
        /// </summary>
        /// <param name="request">Request to process</param>
        public bool ExecuteRequest<T>(PayJunction.Request<T> request) where T : Response
        {
            if (request.IsValid())
            {
                RestSharp.RestClient client = new RestSharp.RestClient(BaseEndpoint);
                RestSharp.RestRequest req = new RestSharp.RestRequest(request.EndPoint, request.Method);
                req.AddHeader("X-PJ-Application-Key", APIKey);
                foreach (string s in request.Headers.Keys)
                {
                    if (!string.IsNullOrWhiteSpace(request.Headers[s]))
                        req.AddHeader(s, request.Headers[s]);
                }

                foreach (string s in request.Params.Keys)
                {
                    if (!string.IsNullOrWhiteSpace(request.Params[s]))
                        req.AddParameter(s, request.Params[s]);
                }

                client.Authenticator = new HttpBasicAuthenticator(UserID, Password);
                IRestResponse resp = client.Execute(req);
                Console.WriteLine("\n\nRaw Json Received\n");
                Console.WriteLine(resp.Content);
                Console.WriteLine("\n\n");
                request.RawResponse = resp.Content;
                request.Response = resp;
                
                return true;
            }

            return false;
        }
    }
}
